### PR TITLE
SdkBundle: Refactor Configuration

### PR DESCRIPTION
**Pull request by tidal**
closes: https://github.com/open-telemetry/opentelemetry-php/issues/652

- Separates config validation/normalization for Resource and Trace
- Abstract away config Validators, Normalizers and Callbacks
- Adds 100% code coverage for config validation/normalization
- Refactors some misc tests

### CODE REVIEW

File: src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/TraceConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    {
+        return true;
+    }
+
+    private static function createSamplerSectionNode(): NodeDefinition
+    {
+        return TreeBuilderHelper::createArrayRootNode(self::SAMPLER_NODE)
+            ->beforeNormalization()
+            ->ifString()
+            ->then(
```
Comment (brettmc) [inline src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/TraceConfiguration.php:68]: this looks like some conditional methods - could it be indented to show that and aide readability? (if cs-fixer will allow it)

File: tests/Unit/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Validator/CustomClassImplementationValidatorTest.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+class CustomClassImplementationValidatorTest extends TestCase
+{
+    public function test_approve(): void
+    {
+        CustomClassImplementationValidator::approve(
+            Exception::class,
+            Throwable::class
+        );
+
+        $this->assertInstanceOf(Throwable::class, new Exception());
```
Comment (brettmc) [inline tests/Unit/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Validator/CustomClassImplementationValidatorTest.php:26]: this looks weird, is it really just a dummy assertion?

File: tests/Unit/Symfony/OtelSdkBundle/Util/ExporterDsnParserTest.php [hunk 3]
```diff
@@ -22,17 +24,17 @@ class ExporterDsnParserTest extends TestCase
         'password' => 'secret',
     ];
 
-    public function testParseFullDsn()
+    public function testParseFullDsn(): void
```
Comment (brettmc) [inline tests/Unit/Symfony/OtelSdkBundle/Util/ExporterDsnParserTest.php:27]: are we missing a cs-fixer rule to enforce snake_case ?

**Review by brettmc — COMMENTED** [review]
lgtm - a couple of questions and suggestions that can be actioned or ignored.
We might need some psalm ignores too, it's not happy.

File: tests/Unit/Symfony/OtelSdkBundle/Util/ExporterDsnParserTest.php (hunk 3 above)
Comment (tidal) [reply tests/Unit/Symfony/OtelSdkBundle/Util/ExporterDsnParserTest.php:27]: Yes, this has never been done in the contrib repo.

**Comment by tidal** [comment]
> lgtm - a couple of questions and suggestions that can be actioned or ignored.
> We might need some psalm ignores too, it's not happy.

I'm working on this. But it will be a bit more than ignores.
The static analysis is really struggling wit this (because the return types dpend on conditions it's not following) , that's why I aready created a lot of those helper methods.
However I want to improve them, as they not really help redability atm.

File: src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/TraceConfiguration.php (hunk 1 above)
Comment (tidal) [reply src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/TraceConfiguration.php:68]: It is, and no sadly CS is always removing my indentations

File: tests/Unit/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Validator/CustomClassImplementationValidatorTest.php (hunk 2 above)
Comment (tidal) [reply tests/Unit/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Validator/CustomClassImplementationValidatorTest.php:26]: This method is throwing an exception, if a class does not implements a certain interface, so it looks like a dummy assertion, but It's not. However the validators do not return anything, these are sinply guard methods.
The downside of `expectNotToPerformAssertions` is also that it removes the code from the coverage report.

File: src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Closure/LogicalEndCallback.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    private static function validate(Closure ...$closure): void
+    {
+        foreach ($closure as $call) {
+            $returnType = (new ReflectionFunction($call))->getReturnType();
+
+            if (!$returnType instanceof ReflectionNamedType || $returnType->getName() !== self::TYPE_BOOL) {
+                throw new InvalidArgumentException(
+                    sprintf(
+                        'All provided closures must have boolean return type. Give: "%s"',
```
Comment (tidal) [inline src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Closure/LogicalEndCallback.php:44]: This is a typo: "Give" -> "Given"

File: src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Closure/MissingAttributesCallback.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                if (!isset($config[$attr])) {
+                    return true;
+                }
+            }
+
+            return false;
+        });
+    }
+
+    public static function createExceptionTrigger(array $requiredAttributes): Closure
```
Comment (tidal) [inline src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Closure/MissingAttributesCallback.php:26]: These two still need to be separated into dedicated Closure factories with a simple *create* methods as in the other one.

File: src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Normalizer/ExporterConfigNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+use OpenTelemetry\Symfony\OtelSdkBundle\DependencyInjection\Configuration\Validator\TypedExporterConfigValidator;
+use OpenTelemetry\Symfony\OtelSdkBundle\DependencyInjection\ConfigurationException;
+use OpenTelemetry\Symfony\OtelSdkBundle\DependencyInjection\ConfigurationInterface;
+use OpenTelemetry\Symfony\OtelSdkBundle\Util\ExporterDsnParser;
+
+class ExporterConfigNormalizer
+{
+    private const ENV_PREFIX = 'env_';
+
+    public static function normalize($config): array
```
Comment (tidal) [inline src/Symfony/OtelSdkBundle/DependencyInjection/Configuration/Normalizer/ExporterConfigNormalizer.php:16]: This method has still has to much cyclomatic complexity.

File: tests/Unit/Symfony/OtelSdkBundle/DependencyInjection/Configuration/ConfigurationBehaviorTest.php [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            }
+
+            public static function getName(): string
+            {
+                return ConfigurationBehaviorTest::CONFIG_NAME;
+            }
+
+            public static function canBeDisabled(): bool
+            {
+                return ConfigurationBehaviorTest::CONFIG_DISABLED;
```
Comment (tidal) [inline tests/Unit/Symfony/OtelSdkBundle/DependencyInjection/Configuration/ConfigurationBehaviorTest.php:131]: `ConfigurationBehaviorTest` should be replaced with `self` here and above

**Pull request closed without merging**